### PR TITLE
Add CSIDriver

### PR DESCRIPTION
**Pull request by jmolmo**
Not sure about several things...

- Should we use a "unit" for the CSIDriver? If the answer is yes, we will need to solve cross reference imports ( "controllers" needs "CSIDriver", and "CSIDriver" needs "controllers", we will have the same problem with other "pkgs".. (Not sure if i understand well the pattern to use the interface for units in the `lvmcluster_controller`)

- Maybe it would be useful a customized `context` with more things .. include the `clientset `can be useful for lot of things

Signed-off-by: Juan Miguel Olmo Martínez <jolmomar@redhat.com>

### CODE REVIEW

**Review by sp98 — CHANGES_REQUESTED** [review]
unit tests?

**Comment by leelavg** [comment]
@jmolmo 

- a lot has to be changed, i'll try to explain the best I can
Some good to know info:
1. Every unit (vgmanager, csicontroller, csinode, csidriver) should implement the interface `resourceManager` to be cleaner for any extensions
2. If any unit doesn't want to reconcile any method (or a specific piece of) it can be left empty (ex. csidriver resource doesn't need an update)


> Should we use a "unit" for the CSIDriver?
1. yes

> "controllers" needs "CSIDriver", and "CSIDriver" needs "controllers"
1. nope, we can get away without deploying CSIDriver resource at all 😅 
2. it's presence only informs k8s that this driver supports/needs these (attachRequired, podInfoOnMount, storageCapacity) functionalities
3. in it's absence it gets the info from one of the controller plugin RPC calls

- Why all this hassle if it's not mandatory?
1. not explaining the detail, we are deploying the CSIDriver just to save a single RPC call and it's roundtrip by setting `attachRequired` to `false`

- Reg PR
1. I was told to use `controllerutil.CreateOrUpdate` for resource creation and utilise controller-runtime client interface https://github.com/kubernetes-sigs/controller-runtime/blob/3ffc700bc2a37c97acd6ce3a03c3526df8a5e489/pkg/client/interfaces.go#L100-L110
2. Owner ref will not work on this resource so main reconciler will set a finalizer on the CR and invokes CSIDriver resource deletion
3. No, we don't need to set `resourceVersion` and it may not work I guess, afaik it's for client & server confirmation that the resource isn't  mutated while we are updating it
4. We are supporting only single CR and better use the logging field available in the main reconciler

Let me know if any of the points doesn't make sense

File: pkg/csi/csidriver.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	csiDriverName = "topolvm.cybozu.com"
+)
+
+var (
+	logger = capnslog.NewPackageLogger("github.com/red-hat-storage/lvm-operator", "csi")
```
Comment (rohantmp) [inline pkg/csi/csidriver.go:17]: let's use the controller-runtime logger? `LVMClusterReconciler.Log`?

File: pkg/util/k8s/k8s.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package k8s
+
+import (
+	"context"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func GetClientSet(ctx context.Context) (*kubernetes.Clientset, error) {
```
Comment (rohantmp) [inline pkg/util/k8s/k8s.go:10]: Let's use the controller-runtime client, it reads from a cache and you supports all runtime.Objects. `LVMClusterReconciler.Client.Create/Update/Delete`

**Review by rohantmp — CHANGES_REQUESTED** [review]
Here is an example from ocs-operator where we took the resourceManager idea from: 
The interface: https://github.com/red-hat-storage/ocs-operator/blob/bf0946e7b4d47b799d209c491f0fcfbd71a4295c/controllers/storagecluster/reconcile.go#L33

A list of implementations: https://github.com/red-hat-storage/ocs-operator/blob/bf0946e7b4d47b799d209c491f0fcfbd71a4295c/controllers/storagecluster/reconcile.go#L352

clients and loggers are available in the first argument of ensureCreated(): https://github.com/red-hat-storage/lvm-operator/blob/a72ebd4a8fb3d478ca6610216264956b3ed52a65/controllers/lvmcluster_controller.go#L126

**Comment by jmolmo** [comment]
> * a lot has to be changed, i'll try to explain the best I can, some good to know info:

Perfectly explained @leelavg , Thank you for take the time to explain!!. I think we can close this PR in favor of your PR (https://github.com/red-hat-storage/lvm-operator/pull/13).

**Comment by jmolmo** [comment]
@rohantmp : Thank you for the multiple references. !! :-)

**Comment by jmolmo** [comment]
Closed in favor of #13

**Pull request closed without merging**